### PR TITLE
Added node aligning options for selected nodes

### PR DESCRIPTION
**Pull request by TeorikDeli**

### CODE REVIEW

**Comment by alelievr** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           

Complexity increasing per file
==============================
- Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/NodeGraphProcessor/NodeGraphProcessor/pullRequest?prid=4513425&bid=15208355)

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs [hunk 1]
```diff
@@ -46,8 +46,24 @@ public class BaseNodeView : NodeView
...
+		
+		private List<Node> selectedNodes = new List<Node>();
+        private float selectedNodesFarLeft;
+        private float selectedNodesNearLeft;
+        private float selectedNodesFarRight;
+        private float selectedNodesNearRight;
+        private float selectedNodesFarTop;
+        private float selectedNodesNearTop;
+        private float selectedNodesFarBottom;
+        private float selectedNodesNearBottom;
```
Comment (alelievr) [inline Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Codacy Issue found: [Remove the field 'selectedNodesNearBottom' and declare it as a local variable in the relevant methods.](https://app.codacy.com/app/alelievr/NodeGraphProcessor/pullRequest?prid=4513425)

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs [hunk 2]
```diff
@@ -234,7 +250,128 @@ public void RemovePort(PortView p)
...
+				if(nodeStyle.top.value.value + nodeHeight < selectedNodesNearBottom) selectedNodesNearBottom = nodeStyle.top.value.value + nodeHeight;
+				if(nodeWidth < selectedNodesShortestWidth) selectedNodesShortestWidth = nodeWidth;
+				if(nodeHeight < selectedNodesShortestHeight) selectedNodesShortestHeight = nodeHeight;
+			}
+
+			selectedNodesAvgHorizontal   = (selectedNodesNearLeft + selectedNodesFarRight) / 2f;
+			selectedNodesAvgVertical    = (selectedNodesNearTop + selectedNodesFarBottom) / 2f;
+		}
+
+		public Rect GetNodeRect(Node node, float left = int.MaxValue, float top = int.MaxValue)
```
Comment (alelievr) [inline Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Codacy Issue found: [Make 'GetNodeRect' a static method.](https://app.codacy.com/app/alelievr/NodeGraphProcessor/pullRequest?prid=4513425)

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs [hunk 3]
```diff
@@ -46,8 +46,24 @@ public class BaseNodeView : NodeView
...
 		[System.NonSerialized]
 		List< IconBadge >						badges = new List< IconBadge >();
-
-			#region  Initialization
+		
+		private List<Node> selectedNodes = new List<Node>();
+        private float selectedNodesFarLeft;
+        private float selectedNodesNearLeft;
+        private float selectedNodesFarRight;
+        private float selectedNodesNearRight;
```
Comment (alelievr) [inline Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Codacy Issue found: [Remove the field 'selectedNodesNearRight' and declare it as a local variable in the relevant methods.](https://app.codacy.com/app/alelievr/NodeGraphProcessor/pullRequest?prid=4513425)

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs [hunk 4]
```diff
@@ -46,8 +46,24 @@ public class BaseNodeView : NodeView
...
+        private float selectedNodesFarTop;
```
Comment (alelievr) [inline Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Codacy Issue found: [Remove the field 'selectedNodesFarTop' and declare it as a local variable in the relevant methods.](https://app.codacy.com/app/alelievr/NodeGraphProcessor/pullRequest?prid=4513425)

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs [hunk 5]
```diff
@@ -46,8 +46,24 @@ public class BaseNodeView : NodeView
...
+        private float selectedNodesNearTop;
+        private float selectedNodesFarBottom;
+        private float selectedNodesNearBottom;
+        private float selectedNodesAvgHorizontal;
+        private float selectedNodesAvgVertical;
+        private float selectedNodesLongestWidth;
+        private float selectedNodesShortestWidth;
+        private float selectedNodesLongestHeight;
```
Comment (alelievr) [inline Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Codacy Issue found: [Remove the field 'selectedNodesLongestHeight' and declare it as a local variable in the relevant methods.](https://app.codacy.com/app/alelievr/NodeGraphProcessor/pullRequest?prid=4513425)

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs [hunk 6]
```diff
@@ -234,7 +250,128 @@ public void RemovePort(PortView p)
...
+			}
+
+			selectedNodesAvgHorizontal   = (selectedNodesNearLeft + selectedNodesFarRight) / 2f;
+			selectedNodesAvgVertical    = (selectedNodesNearTop + selectedNodesFarBottom) / 2f;
+		}
+
+		public Rect GetNodeRect(Node node, float left = int.MaxValue, float top = int.MaxValue)
+		{
+			return new Rect(
+				new Vector2(left != int.MaxValue ? left : node.style.left.value.value, top != int.MaxValue ? top : node.style.top.value.value),
```
Comment (alelievr) [inline Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Codacy Issue found: [Do not check floating point inequality with exact values, use a range instead.](https://app.codacy.com/app/alelievr/NodeGraphProcessor/pullRequest?prid=4513425)

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs [hunk 7]
```diff
@@ -46,8 +46,24 @@ public class BaseNodeView : NodeView
...
+        private float selectedNodesFarRight;
+        private float selectedNodesNearRight;
+        private float selectedNodesFarTop;
+        private float selectedNodesNearTop;
+        private float selectedNodesFarBottom;
+        private float selectedNodesNearBottom;
+        private float selectedNodesAvgHorizontal;
+        private float selectedNodesAvgVertical;
+        private float selectedNodesLongestWidth;
+        private float selectedNodesShortestWidth;
```
Comment (alelievr) [inline Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Codacy Issue found: [Remove the field 'selectedNodesShortestWidth' and declare it as a local variable in the relevant methods.](https://app.codacy.com/app/alelievr/NodeGraphProcessor/pullRequest?prid=4513425)

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs [hunk 8]
```diff
@@ -46,8 +46,24 @@ public class BaseNodeView : NodeView
 
 		[System.NonSerialized]
 		List< IconBadge >						badges = new List< IconBadge >();
-
-			#region  Initialization
+		
+		private List<Node> selectedNodes = new List<Node>();
+        private float selectedNodesFarLeft;
```
Comment (alelievr) [inline Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Codacy Issue found: [Remove the field 'selectedNodesFarLeft' and declare it as a local variable in the relevant methods.](https://app.codacy.com/app/alelievr/NodeGraphProcessor/pullRequest?prid=4513425)

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs [hunk 9]
```diff
@@ -46,8 +46,24 @@ public class BaseNodeView : NodeView
...
+        private float selectedNodesNearLeft;
+        private float selectedNodesFarRight;
+        private float selectedNodesNearRight;
+        private float selectedNodesFarTop;
+        private float selectedNodesNearTop;
+        private float selectedNodesFarBottom;
+        private float selectedNodesNearBottom;
+        private float selectedNodesAvgHorizontal;
+        private float selectedNodesAvgVertical;
+        private float selectedNodesLongestWidth;
```
Comment (alelievr) [inline Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Codacy Issue found: [Remove the field 'selectedNodesLongestWidth' and declare it as a local variable in the relevant methods.](https://app.codacy.com/app/alelievr/NodeGraphProcessor/pullRequest?prid=4513425)

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs [hunk 10]
```diff
@@ -46,8 +46,24 @@ public class BaseNodeView : NodeView
...
+        private float selectedNodesShortestWidth;
+        private float selectedNodesLongestHeight;
+        private float selectedNodesShortestHeight;
```
Comment (alelievr) [inline Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Codacy Issue found: [Remove the field 'selectedNodesShortestHeight' and declare it as a local variable in the relevant methods.](https://app.codacy.com/app/alelievr/NodeGraphProcessor/pullRequest?prid=4513425)

**Comment by alelievr** [comment]
Humm, maybe i'll disable codacy comments ^^ This thing is a bit too verbose

**Comment by TeorikDeli** [comment]
> Humm, maybe i'll disable codacy comments ^^ This thing is a bit too verbose

One thing, I can update GetNodeRect as static, if you want.

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs (hunk 8 above)
Comment (alelievr) [inline Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: do all these fields have to be declared here ? Can they be in the SetValuesForSelectedNodes function ?

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs [hunk 11]
```diff
@@ -413,6 +550,17 @@ public void ChangeLockStatus()
 
         public override void BuildContextualMenu(ContextualMenuPopulateEvent evt)
 		{
+			//	Align
```
Comment (alelievr) [inline Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Can you put these in a separate protected function, so if people want to re-do a contextual menu they can add these options easily ?

Comment (alelievr) [reply Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: I have to do it for the other menu items in this function too :p

**Comment by alelievr** [comment]
> One thing, I can update GetNodeRect as static, if you want.

hum yes, i think it would make sense :)

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs (hunk 8 above)
Comment (TeorikDeli) [reply Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Well, I am using most them in the align functions too (we can reduce some though). I wanted to do "do all calculations, then use whatever needed". I am not really sure how else I can do this, without separating calculations (like, one for to left, one for to right, etc). By the way, if its possible, we can do all this calculations in a OnSelecedNodesChanged kinda event. That way, we can reduce the calling of SetValuesForSelectedNodes function (like change a bool when selection is changed, then when aligning, check that bool to recalculate). IDK, I am open to suggestions 😁

Actually, I am thinking to add another function for setting an equal margin to all selected nodes (which will require, I think, all of the declared fields). I am not sure if I can do more complex aligning options like in Unreal Blueprint, but I can do some of them at least.

By the way, how can I add shortcuts for aligning?

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs (hunk 11 above)
Comment (TeorikDeli) [reply Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Yeap, I can do that 👌🏻

File: Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs (hunk 8 above)
Comment (alelievr) [reply Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs]: Ok, maybe we can keep it like this for now.

To add shortcuts, you can do it here: https://github.com/alelievr/NodeGraphProcessor/blob/76ed7bdea0d65895339c41e545050fa010c723e8/Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseGraphView.cs#L296 I think this still works

**Comment by TeorikDeli** [comment]
Hi 👋🏻

I had been busy for a long time (I couldn't work on our project too!). I will look again where I left 💪🏻 But first, I have to remember what was going on 😂

I see that, you have added so many things. They're looking great 👌🏻

**Comment by alelievr** [comment]
Hi, good to see you back 🙂 

Yes i've been adding some cool stuff to the project, since last time i think there is these new things:
- Now you can override the order of the ports per node
- Change the settings view (now it's a floating panel like shadergraph)
- I implemented the StackNode element from GraphView

**Pull request closed without merging**